### ISSUE TITLE
GP coregionalised regression

### ISSUE THREAD

**Issue by agamemnonc** [SheffieldML/GPy#376]
I am currently following this tutorial for GP (sparse) coregionalised regression:
http://nbviewer.jupyter.org/github/SheffieldML/notebook/blob/master/GPy/coregionalized_regression_tutorial.ipynb

To predict, it seems that one has to transform the input data by adding an extra dimension (output index) as well as specify the model variance. This is all OK and it makes sense when we have different inputs for every output.

However in practice, very often we encounter regression problems with multiple outputs which share exactly the same inputs. Is there a faster / more elegant way of doing inference in that case (right now it seems that a for loop is needed to predict each output)?

**Comment by lawrennd** [comment]
Not just at the moment, but you're right that this special case is a common
one!

We did discuss it a bit but I don't think we reached a particular
conclusion.

**Comment by agamemnonc** [comment]
Thanks Neil.

I now have another comment wrt to coregionalised regression:
I've tried the alternative approach described in the Marathon data notebook. There seems to be a bug when the input is larger than 1D.

Can you please try out the following:

``` python
import numpy as np
import GPy

GPy.util.datasets.authorize_download = lambda x: True # prevents requesting authorization for download.
data = GPy.util.datasets.olympic_sprints()
X = data['X']
y = data['Y']

X = np.hstack((X[:,0][:,None], 0.3*X[:,0][:,None]+np.random.random((np.shape(X)[0],1)), X[:,1][:,None])) # Add another input dimension
kern = GPy.kern.RBF(input_dim = 2, lengthscale=80)**GPy.kern.Coregionalize(input_dim = 2,output_dim=6, rank=6)
model = GPy.models.SparseGPRegression(X, y, kern, num_inducing=100) # This gives an indexing error
```

**Comment by lawrennd** [comment]
The coregionalisation kernel only operates on the output dimension index.
You can select the index using 'active_dims', below I've explicitly set the
coregionalisation active_dim to 2 (which is where the output dimension is
in the model). See if it works for you.

import numpy as np
import GPy
import pods

GPy.util.datasets.authorize_download = lambda x: True # prevents requesting
authorization for download.
data = pods.datasets.olympic_sprints()
X = data['X']
y = data['Y']

X = np.hstack((X[:,0][:,None],
0.3_X[:,0][:,None]+np.random.random((np.shape(X)[0],1)), X[:,1][:,None])) #
Add another input dimension
kern = GPy.kern.RBF(input_dim=2, lengthscale=80, active_dims=[0,
1])_GPy.kern.Coregionalize(1,output_dim=6, active_dims=[2],rank=5)
model = GPy.models.SparseGPRegression(X, y, kern, num_inducing=100)

**Comment by slishak** [comment]
I'm currently having the same problem described above, but the solution proposed by @lawrennd 
doesn't appear to work. Assuming that the `_` should actually be `**`, I can't define the kernel like that:
```python
kern = GPy.kern.RBF(input_dim=2, lengthscale=80, active_dims=[0,1])**GPy.kern.Coregionalize(1,output_dim=6, active_dims=[2],rank=5)
Traceback (most recent call last):
  File "C:\Python27_x64\lib\site-packages\IPython\core\interactiveshell.py", line 2882, in run_code
    exec(code_obj, self.user_global_ns, self.user_ns)
  File "<ipython-input-11-78d8f77bb046>", line 1, in <module>
    kern = GPy.kern.RBF(input_dim=2, lengthscale=80, active_dims=[0,1])**GPy.kern.Coregionalize(1,output_dim=6, active_dims=[2],rank=5)
  File "C:\Python27_x64\lib\site-packages\GPy\kern\src\kern.py", line 314, in __pow__
    assert np.all(other._all_dims_active == range(other.input_dim)), "Can only use kernels, which have their input_dims defined from 0"
AssertionError: Can only use kernels, which have their input_dims defined from 0
```

For info, I'm also working on a problem where I have multiple outputs sharing the same inputs. I'm trying to use `GPy.models.SparseGPCoregionalizedRegression`.

**Comment by lionfish0** [comment]
Hi,
I had some problems with the coregionalisation kernel too - and I didn't understand enough to make a pull-request. As far as I understand Coregionalize will use the 'next free' dimension. So if you've an RBF kernel that uses dims 0 and 1, then the coreg object will use dim 2. I've found active_dims not to work as expected either. I think there's a bug in there somewhere.

As a side note: (terrible hack ahead): I wanted to ignore dim 1 for the kernel I was coregionalising, but use dim 2 as the coreg column. In the end I hacked it by using an RBF on dims 0 /and/ 1, and set the lengthscale of dim 1 to a-very-large-number (so the kernel basically ignored that dimension). Maybe if I get time I'll try to understand the code and make the active_dims thing do what we want, as the status quo is not ideal.

There's a reasonable chance I've misunderstood the workings of this - so if anyone else knows how to get it working, feel free to chip in.
